### PR TITLE
Fix unbound error in ia3.py

### PR DESCRIPTION
**Pull request by His-Wardship**
Fixes an unbound error relating to local variable "result" being referenced prior to being assigned, which arises during fine-tuning using a model loaded in 8-bit, while using Flash Attention (in my case, 2.0.4) and training in BF16 or FP16.

I have no experience nor formal education in computer programming, so apologies if this is redundant or incorrect. Based on the deductive testing the unbound error was caused by loading a model in 8-bit (using Bitsandbytes 41.0) while training in BF16 or FP16 as required by Flash Attention 2. Adding these two lines fixed the unbound error and did not seem to break anything else. The added text is simply a catch-all for where the code does not execute an if-block which assigns a value to result before reaching the current line 551.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thank you a lot for identifying this error. I believe the fix is still not working because the line

```python
if result is None:
```

would still fail with unbound local error. Instead, I _think_ the fix should be:

```python
- result = result * self.ia3_l[self.active_adapter].flatten()
+ result = super().forward(x) * self.ia3_l[self.active_adapter].flatten()
```

but I'd like to ping @SumanthRH to also take a look.

In general, I think the code of `forward` for IA³ (and LoRA too) is too complicated, with a lot of duplication, which makes this error easy to miss. I think we should simplify it.

**Comment by His-Wardship** [comment]
Thank you @BenjaminBossan, my first instinct was to change line 551, as having the same variable on both sides of the equals sign seemed error-prone to me without a catch-all for when the variable was not assigned, but I was not sufficiently familiar with the forward pass logic to make such a change without fear of breaking something else.

In my case, making the change in the PR caused the error to resolve and the IA3 training to properly work on the 8-bit loaded model using Flash Attention (very good training times compared to LoRA, and strong performance after merging the adapter to the model). However, it's entirely possible (and probably likely!) that the PR "fix" just trips something somewhere else and gives the appearance of having fixed the problem.

As I said, I have functionally no knowledge of CS/programming etc., so I have just been reading source code and reasoning it out, so please do dispose of this however is best suited. I apologise for being unable to provide a more fulsome fix/review of the relevant snippet's surroundings.

**Comment by BenjaminBossan** [comment]
> In my case, making the change in the PR caused the error to resolve

Interesting, maybe I'm missing something, but I don't see how that works. Could you please paste the complete error message you got initially?

To me, it seems that the `UnboundLocalError` cannot be resolved by the fix because the check `if result is None` would fail if `result` is not assigned. It would be nice to know what is going on here.

**Comment by His-Wardship** [comment]
Apologies, you are correct; I completely omitted to add that I added an additional line immediately after the line 534 "def forward(self, x: torch.Tensor):", being "result = None". Without this, the PR change is breaking, as you correctly pointed out. I will edit the PR accordingly now.

If I comment out these three lines, after loading a model (in 8-bit) and dataset, loading the PEFT model and config, applying the PEFT model and applying Flash Attention, the following traceback is given. I have also tried this in a few trainers (such as Axolotl) which I have fiddled with to enable IA3 training instead of LoRA, and the same error results if the PR change (with the extra line!) is not made. The references to the folder "language-modeling" are to a local folder which I use to store patches, models and other necessary files. I should have added that I am using the nightly release of PyTorch and CUDA 12.1, but these errors occur on the stable release also, using CUDA 11.7.
 
```
File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/trainer.py:1531, in Trainer.train(self, resume_from_checkpoint, trial, ignore_keys_for_eval, **kwargs)
   1526     self.model_wrapped = self.model
   1528 inner_training_loop = find_executable_batch_size(
   1529     self._inner_training_loop, self._train_batch_size, args.auto_find_batch_size
   1530 )
-> 1531 return inner_training_loop(
   1532     args=args,
   1533     resume_from_checkpoint=resume_from_checkpoint,
   1534     trial=trial,
   1535     ignore_keys_for_eval=ignore_keys_for_eval,
   1536 )

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/trainer.py:1804, in Trainer._inner_training_loop(self, batch_size, args, resume_from_checkpoint, trial, ignore_keys_for_eval)
   1801     self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
   1803 with self.accelerator.accumulate(model):
-> 1804     tr_loss_step = self.training_step(model, inputs)
   1806 if (
   1807     args.logging_nan_inf_filter
   1808     and not is_torch_tpu_available()
   1809     and (torch.isnan(tr_loss_step) or torch.isinf(tr_loss_step))
   1810 ):
   1811     # if loss is nan or inf simply add the average of previous logged losses
   1812     tr_loss += tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/trainer.py:2646, in Trainer.training_step(self, model, inputs)
   2643     return loss_mb.reduce_mean().detach().to(self.args.device)
   2645 with self.compute_loss_context_manager():
-> 2646     loss = self.compute_loss(model, inputs)
   2648 if self.args.n_gpu > 1:
   2649     loss = loss.mean()  # mean() to average on multi-gpu parallel training

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/trainer.py:2671, in Trainer.compute_loss(self, model, inputs, return_outputs)
   2669 else:
   2670     labels = None
-> 2671 outputs = model(**inputs)
   2672 # Save past state if it exists
   2673 # TODO: this needs to be fixed and made cleaner later.
   2674 if self.args.past_index >= 0:

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/utils/operations.py:632, in convert_outputs_to_fp32.<locals>.forward(*args, **kwargs)
    631 def forward(*args, **kwargs):
--> 632     return model_forward(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/utils/operations.py:620, in ConvertOutputsToFp32.__call__(self, *args, **kwargs)
    619 def __call__(self, *args, **kwargs):
--> 620     return convert_to_fp32(self.model_forward(*args, **kwargs))

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/amp/autocast_mode.py:16, in autocast_decorator.<locals>.decorate_autocast(*args, **kwargs)
     13 @functools.wraps(func)
     14 def decorate_autocast(*args, **kwargs):
     15     with autocast_instance:
---> 16         return func(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/peft/peft_model.py:947, in PeftModelForCausalLM.forward(self, input_ids, attention_mask, inputs_embeds, labels, output_attentions, output_hidden_states, return_dict, **kwargs)
    936             raise AssertionError("forward in MPTForCausalLM does not support inputs_embeds")
    937         return self.base_model(
    938             input_ids=input_ids,
    939             attention_mask=attention_mask,
   (...)
    944             **kwargs,
    945         )
--> 947     return self.base_model(
    948         input_ids=input_ids,
    949         attention_mask=attention_mask,
    950         inputs_embeds=inputs_embeds,
    951         labels=labels,
    952         output_attentions=output_attentions,
    953         output_hidden_states=output_hidden_states,
    954         return_dict=return_dict,
    955         **kwargs,
    956     )
    958 batch_size = input_ids.shape[0]
    959 if attention_mask is not None:
    960     # concat prompt attention mask

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/hooks.py:165, in add_hook_to_module.<locals>.new_forward(*args, **kwargs)
    163         output = old_forward(*args, **kwargs)
    164 else:
--> 165     output = old_forward(*args, **kwargs)
    166 return module._hf_hook.post_forward(module, output)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:810, in LlamaForCausalLM.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, labels, use_cache, output_attentions, output_hidden_states, return_dict)
    807 return_dict = return_dict if return_dict is not None else self.config.use_return_dict
    809 # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
--> 810 outputs = self.model(
    811     input_ids=input_ids,
    812     attention_mask=attention_mask,
    813     position_ids=position_ids,
    814     past_key_values=past_key_values,
    815     inputs_embeds=inputs_embeds,
    816     use_cache=use_cache,
    817     output_attentions=output_attentions,
    818     output_hidden_states=output_hidden_states,
    819     return_dict=return_dict,
    820 )
    822 hidden_states = outputs[0]
    823 if self.config.pretraining_tp > 1:

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/hooks.py:165, in add_hook_to_module.<locals>.new_forward(*args, **kwargs)
    163         output = old_forward(*args, **kwargs)
    164 else:
--> 165     output = old_forward(*args, **kwargs)
    166 return module._hf_hook.post_forward(module, output)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:690, in LlamaModel.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, use_cache, output_attentions, output_hidden_states, return_dict)
    686             return module(*inputs, output_attentions, None)
    688         return custom_forward
--> 690     layer_outputs = torch.utils.checkpoint.checkpoint(
    691         create_custom_forward(decoder_layer),
    692         hidden_states,
    693         attention_mask,
    694         position_ids,
    695         None,
    696     )
    697 else:
    698     layer_outputs = decoder_layer(
    699         hidden_states,
    700         attention_mask=attention_mask,
   (...)
    704         use_cache=use_cache,
    705     )

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/_compile.py:24, in _disable_dynamo.<locals>.inner(*args, **kwargs)
     20 @functools.wraps(fn)
     21 def inner(*args, **kwargs):
     22     import torch._dynamo
---> 24     return torch._dynamo.disable(fn, recursive)(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:321, in _TorchDynamoContext.__call__.<locals>._fn(*args, **kwargs)
    319 dynamic_ctx.__enter__()
    320 try:
--> 321     return fn(*args, **kwargs)
    322 finally:
    323     set_eval_frame(prior)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/_dynamo/external_utils.py:17, in wrap_inline.<locals>.inner(*args, **kwargs)
     15 @functools.wraps(fn)
     16 def inner(*args, **kwargs):
---> 17     return fn(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/utils/checkpoint.py:450, in checkpoint(function, use_reentrant, context_fn, determinism_check, debug, *args, **kwargs)
    445     if context_fn is not noop_context_fn or debug is not False:
    446         raise ValueError(
    447             "Passing `context_fn` or `debug` is only supported when "
    448             "use_reentrant=False."
    449         )
--> 450     return CheckpointFunction.apply(function, preserve, *args)
    451 else:
    452     gen = _checkpoint_without_reentrant_generator(
    453         function, preserve, context_fn, determinism_check, debug, *args, **kwargs
    454     )

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/autograd/function.py:539, in Function.apply(cls, *args, **kwargs)
    536 if not torch._C._are_functorch_transforms_active():
    537     # See NOTE: [functorch vjp and autograd interaction]
    538     args = _functorch.utils.unwrap_dead_wrappers(args)
--> 539     return super().apply(*args, **kwargs)  # type: ignore[misc]
    541 if cls.setup_context == _SingleLevelFunction.setup_context:
    542     raise RuntimeError(
    543         "In order to use an autograd.Function with functorch transforms "
    544         "(vmap, grad, jvp, jacrev, ...), it must override the setup_context "
    545         "staticmethod. For more details, please see "
    546         "https://pytorch.org/docs/master/notes/extending.func.html"
    547     )

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/utils/checkpoint.py:230, in CheckpointFunction.forward(ctx, run_function, preserve_rng_state, *args)
    227 ctx.save_for_backward(*tensor_inputs)
    229 with torch.no_grad():
--> 230     outputs = run_function(*args)
    231 return outputs

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:686, in LlamaModel.forward.<locals>.create_custom_forward.<locals>.custom_forward(*inputs)
    684 def custom_forward(*inputs):
    685     # None for past_key_value
--> 686     return module(*inputs, output_attentions, None)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/hooks.py:165, in add_hook_to_module.<locals>.new_forward(*args, **kwargs)
    163         output = old_forward(*args, **kwargs)
    164 else:
--> 165     output = old_forward(*args, **kwargs)
    166 return module._hf_hook.post_forward(module, output)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py:413, in LlamaDecoderLayer.forward(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache)
    410 hidden_states = self.input_layernorm(hidden_states)
    412 # Self Attention
--> 413 hidden_states, self_attn_weights, present_key_value = self.self_attn(
    414     hidden_states=hidden_states,
    415     attention_mask=attention_mask,
    416     position_ids=position_ids,
    417     past_key_value=past_key_value,
    418     output_attentions=output_attentions,
    419     use_cache=use_cache,
    420 )
    421 hidden_states = residual + hidden_states
    423 # Fully Connected

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/accelerate/hooks.py:165, in add_hook_to_module.<locals>.new_forward(*args, **kwargs)
    163         output = old_forward(*args, **kwargs)
    164 else:
--> 165     output = old_forward(*args, **kwargs)
    166 return module._hf_hook.post_forward(module, output)

File ~/language-modeling/llama_patch.py:46, in forward(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache)
     43 bsz, q_len, _ = hidden_states.size()
     45 query_states = self.q_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
---> 46 key_states = self.k_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
     47 value_states = self.v_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
     48 # [bsz, q_len, nh, hd]
     49 # [bsz, nh, q_len, hd]

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File ~/anaconda3/envs/cuda12/lib/python3.10/site-packages/peft/tuners/ia3.py:554, in Linear8bitLt.forward(self, x)
    550             result = super().forward(x * self.ia3_l[self.active_adapter].flatten())
    551         else:
    552             #if result is None:
    553             #    result = super().forward(x)
--> 554             result = result * self.ia3_l[self.active_adapter].flatten()
    555 return result
```
Apologies again for the oversight, I should have done an actual doc-to-doc comparison rather than just trying to recall the changes I made.

**Comment by BenjaminBossan** [comment]
> I completely omitted to add that I added an additional line immediately after the line 534 "def forward(self, x: torch.Tensor):", being "result = None"

Ah yes, that makes sense. Still, I think that my one-line fix provided above should also work and doesn't require to reassign the `result` variable, maybe you can test it.

Overall, I think, however, that the whole `forward` method could be simplified:

```python
        # current forward (incl. bugfix)
        def forward(self, x: torch.Tensor):
            if self.disable_adapters or self.active_adapter not in self.ia3_l.keys():
                return super().forward(x)
            else:
                if not torch.is_autocast_enabled():
                    if x.dtype != torch.float32:
                        x = x.float()
                    if self.is_feedforward:
                        result = super().forward(x * self.ia3_l[self.active_adapter].flatten())
                    else:
                        result = super().forward(x)
                        expected_dtype = result.dtype
                        result = (result * self.ia3_l[self.active_adapter].flatten()).to(expected_dtype)
                else:
                    if self.is_feedforward:
                        result = super().forward(x * self.ia3_l[self.active_adapter].flatten())
                    else:
                        result = super().forward(x)
                        result = result * self.ia3_l[self.active_adapter].flatten()
            return result

        # proposed new forward (incl. bugfix)
        def forward(self, x: torch.Tensor):
            if self.disable_adapters or (self.active_adapter not in self.ia3_l.keys()):
                return super().forward(x)

            requires_conversion = (not torch.is_autocast_enabled()) and (x.dtype != torch.float32)
            if requires_conversion:
                x = x.float()

            ia3_scaling = self.ia3_l[self.active_adapter].flatten()
            if self.is_feedforward:
                result = super().forward(x * ia3_scaling)
            else:
                result = super().forward(x)
                expected_dtype = result.dtype
                result = result * ia3_scaling

            if requires_conversion:
                result = result.to(expected_dtype)
            return result
```

If I'm not missing something, this code should be exactly the same as the existing code (+ bugfix) but only ever has one level of nesting instead of three, and has no duplication, so it should be less error prone. I'm not sure if it's correct that the dtype-conversion is _not_ applied when `self.is_feedforward`, but that's already the case with the existing code.

Of course, similar changes should be made to the non-8bit layer and to the corresponding (Ada)LoRA layers.

**Comment by His-Wardship** [comment]
I have inserted the proposed code into ia3.py and it works as expected, having tested on a full training run on a small dataset (using Llama-2-7B). Not sure what the etiquette is here, should I edit these changes into the PR?

**Comment by BenjaminBossan** [comment]
Thanks for testing.

> I have inserted the proposed code into ia3.py and it works as expected

Do you mean the one-liner or the whole rewrite of the method?

> Not sure what the etiquette is here, should I edit these changes into the PR?

Yes, you can do that. But let's first wait for a second opinion (ping @pacman100).

**Comment by His-Wardship** [comment]
I mean the whole rewrite which you proposed under the title: 
        # proposed new forward (incl. bugfix)

Thank you for your help! I shall wait on @pacman100. I imagine there will be more interest in IA3, and more attention to potential bugs in its surrounding code, once the more popular UI repos (text-generation-webui, axolotl etc.) begin to support it, as it seems from my testing to be far superior to LoRA. Of course, the major catch at the moment is that it only supports as low as 8-bit, so if you wanted to train an IA3 adapter on a 33B model, you would need 2x 24GB cards or an industrial-grade card like an A100, which I imagine rather dampens interest.

**Comment by pacman100** [comment]
Hello @BenjaminBossan and @His-Wardship, I really like the simplified code that Benjamin has shared above, all in for simplifying code to be more readable and maintainable. To quote PyTorch Design principle `Simple is better than complex`

**Comment by pacman100** [comment]
@His-Wardship, could you please share experimental results comparing IA^3 and LoRA so that it helps community to further grow interest in IA^3

**Comment by His-Wardship** [comment]
@BenjaminBossan Thank you very much. Based on @pacman100's comments, I have replaced my adjustment in the PR to use your proposed code block instead, and committed those changes. 

I will happily try to produce some more fulsome comparison data between IA3 and LoRA (my own tests for determining whether to use LoRA or IA3 were very "back of a napkin" and wouldn't be considered rigorous). Once my current training process finishes this evening, I shall rerun some small samples to compare training times, performance, resource requirements etc.

File: src/peft/tuners/ia3.py
Comment (BenjaminBossan) [inline src/peft/tuners/ia3.py:551]:
```suggestion
                if requires_conversion:
                    result = result.to(expected_dtype)
```

Oh, I see a bug in my proposed code. `expected_dtype` might not exist, so this either needs to be indented (as suggested), leading to 2 levels of indentation ;-( or we would have to do something like:

```python
# near the top of the function
expected_dtype = None

...

if requires_conversion and expected_dtype is not None:
    result = result.to(expected_dtype)
```

However, there is still this issue I mentioned earlier:

> I'm not sure if it's correct that the dtype-conversion is not applied when `self.is_feedforward`, but that's already the case with the existing code.

@SumanthRH @pacman100

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thx @His-Wardship for making the adjustment. There are other functions that could be cleaned up in a similar manner, but we can do that in a separate PR.

There is only one small issue (from my proposed code) that I found, which should be fixed before merging.

**Comment by His-Wardship** [comment]
@BenjaminBossan Thank you! While I lack the technical knowledge to properly consider the change, after testing it locally and finding (unsurprisingly!) that it was not breaking, I'll commit your change on your own recommendation.

With respect to the comment relating to dtype-conversion, am I correct in thinking you are describing something similar to the snippet below? As I didn't fully appreciate the effect of the indentation, I just initialized a value for expected_dtype prior to the relevant if-block and moved the conversion code outside of the else-block.

```
def forward(self, x: torch.Tensor):
    if self.disable_adapters or (self.active_adapter not in self.ia3_l.keys()):
        return super().forward(x)

    requires_conversion = (not torch.is_autocast_enabled()) and (x.dtype != torch.float32)
    if requires_conversion:
        x = x.float()

    ia3_scaling = self.ia3_l[self.active_adapter].flatten()
    expected_dtype = x.dtype 
    if self.is_feedforward:
        result = super().forward(x * ia3_scaling)
    else:
        result = super().forward(x)
        expected_dtype = result.dtype
        result = result * ia3_scaling

    if requires_conversion:
        result = result.to(expected_dtype)
            
    return result

```

**Comment by BenjaminBossan** [comment]
> With respect to the comment relating to dtype-conversion, am I correct in thinking you are describing something similar to the snippet below?

With the proposed change, if `self.is_feedforward`, the result would be cast to the the same dtype as the input `x` after conversion. As I don't know why the casting to `expected_dtype` was required in the first place, I can't say if the change is correct. To check further, I modified the test suite to use 8bit and ran it for IA³ -- in the tests, the dtype conversion was never necessary. Ideally, @SumanthRH could comment on that. If we don't hear back, it's probably better to leave the code as is.

**Review by pacman100 — APPROVED** [review]
Hello @His-Wardship, Thank you and @BenjaminBossan for working on this, LGTM! 🚀

@BenjaminBossan, wrt dtype conversion, @younesbelkada worked on it for LoRA and should be able to provide more details on it. In short, when autocasting is disabled, we are converting the input to FP32 as the adapter weights are in FP32 to avoid dtype mismatch errors.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @His-Wardship 
This looks great to me, yes I second what @pacman100 said, with respect to casting, this is needed for half-precision / quantized models in case autocast is not enabled!

File: src/peft/tuners/ia3.py
Comment (pacman100) [inline src/peft/tuners/ia3.py:551]:
```suggestion
            if requires_conversion:
                result = result.to(expected_dtype)
```

**Review by pacman100 — COMMENTED** [review]
Hello, the result should be converted back to expected dtype in both cases. Added a suggestion. Thanks to Benjamin for pointing this out

**Pull request merged**
